File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def InOrder(self, root, k):
        if root == None:
            return

        self.InOrder(root.left, k)

        self.count+= 1
        if self.count == k:
            self.kthvalue = root.val
            return
        
        self.InOrder(root.right, k)

        
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        
        self.count = 0
        self.InOrder(root, k)
        return self.kthvalue

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int dfs(TreeNode* root, int &ans, int maxsum) {

        if (root == NULL) return 0;

        int leftsum = max(dfs(root->left, ans, maxsum), 0);
        int rightsum = max(dfs(root->right, ans, maxsum), 0);

        
        int throughRoot = root->val + leftsum + rightsum;
        ans = max(ans, throughRoot);

        maxsum = root->val + max(leftsum, rightsum);
        return maxsum;        
    }
    int maxPathSum(TreeNode* root) {

        if (root == NULL) return 0;

        int ans = INT_MIN;
        int maxsum = 0;

        dfs(root, ans, maxsum);

        return ans;
    }
};
